### ISSUE TITLE
grpc.experimental.aio: returning an asynchronous generator from a function defined with def

### ISSUE THREAD

**Issue by vitamins** [grpc/grpc#24603]
### What version of gRPC and what language are you using?
version of grpc 1.33.0-1
language Python

### What operating system (Linux, Windows,...) and version?
Linux, Arch Linux

### What runtime / compiler are you using (e.g. python version or version of gcc)
Package versions
grpc 1.33.0-1
python-grpcio 1.33.0-1

### What did you do?

Define an rpc in a .proto file:

` rpc AnRPC() returns (stream Something) {}`

Implement a servicer class

Variant A: Returning an asynchronous generator from a function defined with def does raise an exception.
```
import grpc
from grpc.experimental import aio

async def an_async_generator():
    for x in range(10):
        yield a_service_pb2.Something()

class AServiceServicer(a_service_pb2_grpc.AServiceServicer):

    def AnRPC(self, request, context):
        # returns an async generator
        logging.info(f'AnRPC: Received request.')
        return an_async_generator()
```

Variant B: Yielding from a function defined with async def does work o.k.
```
class AServiceServicer(a_service_pb2_grpc.AServiceServicer):

    async def AnRPC(self, request, context):
        # returns an async generator
        logging.info(f'AnRPC: Received request.')
        async for x in an_async_generator():
            yield x
```

### What did you expect to see?
Both variants A and B should work equivalently in python.

### What did you see instead?
This is the exception raised by variant A.
```
INFO:root:GetCharInfoStream: Received request.
ERROR:grpc._cython.cygrpc:Unexpected [TypeError] raised by servicer method [/AService/AnRCP]
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 646, in grpc._cython.cygrpc._handle_exceptions
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 752, in _handle_rpc
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 544, in _handle_unary_stream_rpc
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 459, in _finish_handler_with_stream_responses
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 169, in generator_to_async_generator
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 151, in grpc._cython.cygrpc.generator_to_async_generator.yield_to_queue
TypeError: 'async_generator' object is not iterable
```

**Comment by lidizheng** [comment]
This is strange. We have test cases guarding this particular use case. I tested it with our route guide example, and it works fine.

```python
class RouteGuideServicer(route_guide_pb2_grpc.RouteGuideServicer):

    async def RouteChat(self, request_iterator, context):
        prev_notes = []
        async for new_note in request_iterator:
            for prev_note in prev_notes:
                if prev_note.location == new_note.location:
                    yield prev_note
            prev_notes.append(new_note)

async def serve():
    server = grpc.aio.server()
    route_guide_pb2_grpc.add_RouteGuideServicer_to_server(
        RouteGuideServicer(), server)
    server.add_insecure_port('[::]:50051')
    await server.start()
    await server.wait_for_termination()
```

Maybe we do have a bug in an uncovered path, can you share a complete reproduction? @vitamins

**Comment by rschmied** [comment]
I can see the same issue with the code I'm trying to put together.  The difference between my code (and I think @vitamins code) is that we have the generator on the server side whereas your sample code consumes a generator from the client.  

Apparently, from the exception @vitamins provided, the generator is already async but then there's `generator_to_async_generator` which tries to wrap a non-async generator into a future.

**Comment by lidizheng** [comment]
As the reporter was inactive to respond, I will go ahead and close this issue. If similar bug appears, please try to file another issue with detailed error log and reproduction case.

**Comment by mickours** [comment]
I've experience the same issue  with an async generator function not directly in the RPC method (Variant A) throwing the same error.

I finally make works by calling the external function async generator in a loop. If I take the previous example here is a working version:
```python3
import grpc
from grpc.experimental import aio

async def an_async_generator():
    for x in range(10):
        yield a_service_pb2.Something()

class AServiceServicer(a_service_pb2_grpc.AServiceServicer):

    def AnRPC(self, request, context):
        # returns an async generator
        logging.info(f'AnRPC: Received request.')
        async for msg in  an_async_generator():
            yield msg
```

**Comment by lidizheng** [comment]
@mickours Can you open a new issue with your environment details and a reproduce snippet? Also, we just released v1.40.0 last week, can you double check if the problem persist in the latest version?

**Comment by mickours** [comment]
In the end I think that this is not an issue of grpc: yielding from a generator is not the same as returning a generator.

So, what is missing here is just a good example so I've made a PR to add it: #27343
